virttest/virt_vm.py

```diff
@@ -1049,66 +1049,69 @@ def wait_for_login(self, nic_index=0, timeout=LOGIN_WAIT_TIMEOUT,
 
+        error = None
         logging.debug("Attempting to log into '%s' (timeout %ds)",
                       self.name, timeout)
-        start = time.time()
+        start_time = time.time()
         try:
             self.wait_for_get_address(nic_index,
                                       timeout=timeout,
                                       ip_version=self.ip_version)
-        except Exception:
+        except Exception, e:
             self.verify_alive()
             print_guest_network_info()
             if not (serial or restart_network):
                 raise
-            if restart_network:
-                serial_login_args = (timeout, internal_timeout,
-                                     restart_network, username,
-                                     password, False)
-                session = self.wait_for_serial_login(*serial_login_args)
-                if serial:
-                    return session
-                session.close()
+            session = self.wait_for_serial_login(timeout, internal_timeout,
+                                                 restart_network, username,
+                                                 password, False)
+            if serial:
+                return session
+            session.close()
+            error = e
 
-        end = time.time()
-        # try to login if VM bootup really
-        timeout -= (end - start)
-        login_timeout = time.time() + timeout
-        while True:
+        # try to login if VM bootup really, at least once
+        not_tried = True
+        end_time = start_time + timeout
+        while time.time() < end_time or not_tried:
             try:
-                return self.login(nic_index,
-                                  internal_timeout,
-                                  username,
-                                  password)
-            except Exception:
-                pass
-            # Re-try login via serial console, if seiral is true
-            if serial:
-                try:
-                    return self.serial_login(internal_timeout,
-                                             username,
-                                             password,
-                                             False)
-                except Exception:
-                    pass
-            if time.time() > login_timeout:
-                print_guest_network_info()
-                raise remote.LoginTimeoutError('exceeded %s s timeout' %
-                                               login_timeout)
+                return self.login(nic_index, internal_timeout,
+                                  username, password)
+            except (remote.LoginAuthenticationError,
+                    remote.LoginBadClientError):
+                if serial:
+                    break
+                raise
+            except Exception, e:
+                error = e
+            not_tried = False
+
+        print_guest_network_info()
+        if serial:
```

Comment:
if serial is "True", It will return in line 1088
```
if serial:
    return session
```
So is the "if segment" there should be removed ?

Reply:
No, L1088 would be executed only if error occurred while running `wait_for_get_address`, so here must check again.